### PR TITLE
HPCC-29969 Add connection retry logic to vault access code

### PR DESCRIPTION
**Pull request by afishbeck**
## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: helm/hpcc/templates/_helpers.tpl [hunk 1]
```diff
@@ -604,6 +604,21 @@ vaults:
     {{- if index $vault "appRoleSecret" }}
       appRoleSecret: {{ index $vault "appRoleSecret" }}
     {{- end -}}
+    {{- if (hasKey $vault "retries") }}
+      retries: {{ $vault.retries }}
+    {{- end }}
+    {{- if (hasKey $vault "retry_wait_seconds") }}
+      retry_wait_seconds: {{ $vault.retry_wait_seconds }}
```
Comment (ghalliday) [inline helm/hpcc/templates/_helpers.tpl:611]: Minor.  There seems to be a mixture of naming conventions for compound words:

verify_server
client-secret
appRoleSecret

which is confusing.  I think we have used camelCase as a standard nearly everywhere else - so these should be consistent with that.

File: system/jlib/jsecrets.cpp [hunk 2]
```diff
@@ -350,7 +361,7 @@ class CVault
                 StringBuffer tokenText;
                 if (getSecretKeyValue(clientToken, clientSecret, "token"))
                 {
-                    authType = VaultAuthType::appRole;
+                    authType = VaultAuthType::token;
```
Comment (ghalliday) [inline system/jlib/jsecrets.cpp:364]: Is this a separate bug fix?  Should it be a separate PR, or at least explicitly mentioned in the commit.  (Or did it have no effect?)

File: system/jlib/jsecrets.cpp [hunk 3]
```diff
@@ -306,6 +306,11 @@ class CVault
     time_t clientTokenExpiration = 0;
     bool clientTokenRenewable = false;
     bool verify_server = true;
+    unsigned retries = 3;
+    time_t retry_wait_seconds = 1;
```
Comment (ghalliday) [inline system/jlib/jsecrets.cpp:310]: similar comment about variable names - normally we use camel case.

**Review by ghalliday — CHANGES_REQUESTED** [review]
Looks like a good change.  Main comments are about naming conventions.

File: system/jlib/jsecrets.cpp (hunk 3 above)
Comment (afishbeck) [reply system/jlib/jsecrets.cpp:310]: It was broken, but I doubt this is currently used by anyone but me, although it could be useful at some point.  I'll create a separate JIRA/PR.

File: system/jlib/jsecrets.cpp (hunk 2 above)
Comment (afishbeck) [reply system/jlib/jsecrets.cpp:364]: It was broken, but I doubt this is currently used by anyone but me, although it could be useful at some point.  I'll create a separate JIRA/PR.

**Comment by afishbeck** [comment]
@ghalliday updated based on review comments.

File: system/jlib/jsecrets.cpp [hunk 4]
```diff
@@ -438,6 +449,22 @@ class CVault
 
     CVaultKind getVaultKind() const { return kind; }
 
+    void initClient(httplib::Client &cli, httplib::Headers &headers, unsigned &numRetries)
+    {
+        numRetries = retries;
+        cli.enable_server_certificate_verification(verify_server);
+        if (connectTimeout)
```
Comment (ghalliday) [inline system/jlib/jsecrets.cpp:456]: Are seconds the correct granularity for setting timeouts?  Should it be using us instead?

**Review by ghalliday — COMMENTED** [review]
@afishbeck looks good.  Please squash.  One question about using us v s for the timeout granularity.

**Comment by afishbeck** [comment]
@ghalliday updated the new timeout values to milliseconds which is more consistent with existing timeouts and more useful.

**Comment by afishbeck** [comment]
@ghalliday shall I go ahead and squash, and then you can do a quick re-review before taking action?  The recent change was using MS for timeouts, and therefor converting to the timeval that is used by httplib.h.

**Comment by crosgithub** [comment]
When do we expect a fix for the timing issue?  I have a health process that kicks off 3 WUs every day on on-prem dataland.  The first WU always fails with 

1 error(s), 0 warning(s)
eclcc: (0,0): Error C3000: Vault [my-git-vault] approle auth error missing login response, error 4

https://boca-dataland-dev.hpcc.risk.regn.net:18012/esp/files/stub%2Ehtm?Wuid=W20230811-060002&Widget=WUDetailsWidget#/stub/Summary

This issue started on 20230802.

The 2 subsequent WUs generally always complete.

File: system/jlib/jsecrets.cpp [hunk 5]
```diff
@@ -350,7 +378,7 @@ class CVault
                 StringBuffer tokenText;
                 if (getSecretKeyValue(clientToken, clientSecret, "token"))
                 {
-                    authType = VaultAuthType::token;
+                    authType = VaultAuthType::appRole;
```
Comment (ghalliday) [inline system/jlib/jsecrets.cpp:381]: This looks like some old code sneaked back in.

Comment (afishbeck) [reply system/jlib/jsecrets.cpp]: @ghalliday fixed.

**Review by ghalliday — APPROVED** [review]
